metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -293,12 +302,12 @@ rd_succ_bv_type route_at_forks(const graph::DBGSuccinct &graph,
             [&](int32_t count) {
                 // TODO: skip single outgoing
                 outgoing_counts.push_back(count);
-                if (last[graph.kmer_to_boss_index(graph_idx)]) {
+                if (last[from_graph_index(graph, graph_idx)]) {
                     // pick the node with the largest count
                     size_t max_pos = std::max_element(outgoing_counts.rbegin(),
                                                       outgoing_counts.rend())
                                      - outgoing_counts.rbegin();
-                    rd_succ_bv[graph.kmer_to_boss_index(graph_idx - max_pos)] = true;
+                    rd_succ_bv[from_graph_index(graph, graph_idx - max_pos)] = true;
                     outgoing_counts.resize(0);
                 }
                 graph_idx++;
```

Comment:
GitHub doesn't allow me to put comments on unedited sections :slightly_frowning_face: 

But essentially, changing `route_at_forks` is somewhat straightforward, but `build_pred_succ` and to a smaller extent `assign_anchors` appear to depend more heavily on the stuff from `DBGSuccinct`.

For example, instead of
```cpp
    graph::DBGSuccinct graph(2);
    logger->trace("Loading graph...");
    if (!graph.load(graph_fname)) {
        logger->error("Cannot load graph from {}", graph_fname);
        std::exit(1);
    }
```

What actual class should I try to load? Should it be a template parameter to `build_pred_succ` and `assign_anchors`? Also can you tell me a bit more on a high level what `build_pred_succ` and its BOSS-specific part in particular are supposed to do?

Reply:
I think it would make more sense to modify these functions so they take `const DeBruijnGraph&` as an argument instead of the path to the graph. There's a loading function in `cli/load/load_graph.hpp`, but we'd have to do more refactoring to make that accessible to this function.

`build_pred_succ` is supposed to create predecessor and successor arrays, so for each node index, they store the index of the first predecessor and last successor (respectively). "First" and "last" are determined lexicographically

Reply:
In the current version of the function, it accepts an optional bit vector `rd_succ` which marks nodes that are successors. If this vector exists, then the index of the successor of a node for which `rd_succ` stores `true` is used instead of the last successor

Reply:
For predecessors, the `bwd` + `call_incoming_to_target` methods can be replaced with `DeBruijnGraph::adjacent_incoming_nodes`